File: examples/native-modules/cpp/src/ping.cpp
// Copyright 2026 Dimensional Inc.
// SPDX-License-Identifier: Apache-2.0

#include <chrono>
#include <cstdint>
#include <thread>

#include "dimos/native.hpp"
#include "geometry_msgs/Twist.hpp"

using dimos::native::Builder;
using dimos::native::Config;
using dimos::native::Module;
using dimos::native::Output;
namespace logging = dimos::native::log;
using geometry_msgs::Twist;

constexpr std::chrono::milliseconds kPublishPeriod{200};

class Ping : public Module {
public:
    void build(Builder& builder, Config& /*config*/) override {
        // publish data topic
        data_ = builder.output<Twist>("data");

        // input confirm topic
        builder.input<Twist>("confirm", &Ping::on_confirm, this);
    }

    void setup() override {
        producer_ = std::thread([this] {
            std::uint64_t seq = 0;
            while (!shutdown_requested()) {
                Twist msg;
                msg.linear.x = static_cast<double>(seq);
                msg.linear.y = 0.0;
                msg.linear.z = 0.0;
                msg.angular.x = 0.0;
                msg.angular.y = 0.0;
                msg.angular.z = 0.0;
                data_.publish(msg);
                ++seq;
                std::this_thread::sleep_for(kPublishPeriod);
            }
        });
    }

    void teardown() override {
        if (producer_.joinable()) {
            producer_.join();
        }
    }

private:
    void on_confirm(const Twist& echo) {
        logging::info("echo received (cpp)",
                      {logging::Field("seq", static_cast<std::int64_t>(echo.linear.x)),
                       logging::Field("sample_config",
                                      static_cast<std::int64_t>(echo.angular.z))});
    }

    Output<Twist> data_;
    std::thread producer_;
};

int main() {
    dimos::native::run_with_transport<Ping>();
    return 0;
}

File: examples/native-modules/rust/src/ping.rs
// Copyright 2026 Dimensional Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use dimos_module::{run_with_transport, Input, Module, Output};
use lcm_msgs::geometry_msgs::{Twist, Vector3};
use tokio::time::{interval, Duration};

#[derive(Module)]
#[module(setup = start_publisher)]
struct Ping {
    #[input(decode = Twist::decode)]
    confirm: Input<Twist>,

    #[output(encode = Twist::encode)]
    data: Output<Twist>,
}

impl Ping {
    async fn start_publisher(&mut self) {
        let data = self.data.clone();
        tokio::spawn(async move {
            let mut ticker = interval(Duration::from_millis(200));
            let mut seq = 0u64;
            loop {
                ticker.tick().await;
                let msg = Twist {
                    linear: Vector3 {
                        x: seq as f64,
                        y: 0.0,
                        z: 0.0,
                    },
                    angular: Vector3 {
                        x: 0.0,
                        y: 0.0,
                        z: 0.0,
                    },
                };
                data.publish(&msg).await.ok();
                seq += 1;
            }
        });
    }

    async fn handle_confirm(&mut self, echo: Twist) {
        tracing::info!(
            seq = echo.linear.x as u64,
            sample_config = echo.angular.z as i64,
            "echo received (rust)",
        );
    }
}

#[tokio::main]
async fn main() {
    run_with_transport::<Ping>().await;
}
